src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+                // the end of the list.
+                if (!parameterSymbol.HasExplicitDefaultValue && !parameterSymbol.IsParams &&
+                        !parameterSymbol.Type.Equals(compilation.GetTypeByMetadataName(typeof(CancellationToken)?.FullName!)))
+                {
+                    insertionIndex++;
+                }
+            }
+
+            return insertionIndex;
+        }
+
+        /// <summary>
+        /// For the trampoline case, it goes through the invocations and adds an argument which is a 
+        /// call to the extracted method.
+        /// Introduces a new method overload or new trampoline method.
+        /// Updates the original method site with a newly introduced parameter.
+        /// </summary>
+        private static async Task<Solution> ModifyDocumentInvocationsTrampolineAndIntroduceParameterAsync(
+            Compilation compilation, Solution modifiedSolution, Document currentDocument, Document originalDocument,
+            ImmutableArray<TInvocationExpressionSyntax> invocations, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+            bool allOccurrences, string parameterName, TExpressionSyntax expression, bool trampoline, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(currentDocument);
+            var semanticFacts = currentDocument.GetRequiredLanguageService<ISemanticFactsService>();
+            var invocationSemanticModel = await currentDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var root = await currentDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, generator);
+            var syntaxFacts = currentDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+            var insertionIndex = GetInsertionIndex(compilation, methodSymbol, syntaxFacts, containingMethod);
+            var newMethodIdentifier = methodSymbol.Name + "_" + parameterName;
+
+            if (trampoline)
+            {
+                foreach (var invocationExpression in invocations)
+                {
+                    editor.ReplaceNode(invocationExpression, (currentInvocation, _) =>
+                    {
+                        return GenerateNewInvocationExpressionForTrampoline(syntaxFacts, editor, currentInvocation,
+                            invocationExpression, newMethodIdentifier, insertionIndex);
+                    });
+                }
+            }
+
+            // If you are at the original document, then also introduce the new method and introduce the parameter.
+            if (currentDocument.Id == originalDocument.Id)
+            {
+                if (trampoline)
+                {
+                    await GenerateNewMethodAsync(originalDocument, expression, methodSymbol, containingMethod, newMethodIdentifier,
```

Comment:
this name should have trampoline in it to clearly indicate what it is generating.  'new method' is very vague.

Reply:
do your generation methods ensure that you are generating next to the current method?

Reply:
They always insert before the current method.